File: Sources/SSHTunnelManager/Models/ExampleProfiles.swift
import Foundation

/// Example profiles seeded on first launch so a new user can see how different
/// options come together. They use `example.com` placeholder hosts (which won't
/// resolve) and are safe to edit or delete — see `ProfileStore` for the one-time
/// seeding logic.
extension SSHProfile {
    static var examples: [SSHProfile] {
        [
            databaseTunnelExample,
            socksProxyExample,
            bastionShellExample,
            remoteShareExample,
        ]
    }

    /// `-L` local forward, tunnels-only (`-N`): reach a remote database on localhost.
    private static var databaseTunnelExample: SSHProfile {
        var p = SSHProfile()
        p.name = "Example: Database tunnel (-L)"
        p.host = "db.internal.example.com"
        p.username = "deploy"
        p.openShell = false          // tunnels only → ssh runs with -N
        p.keepAlive = true
        p.theme = "ocean"
        p.forwards = [
            PortForward(type: .local,
                        listenPort: "5433",
                        targetHost: "127.0.0.1",
                        targetPort: "5432"),
        ]
        p.snippets = [
            CommandSnippet(label: "Connect with psql",
                           command: #"psql "host=127.0.0.1 port=5433 dbname=app user=app""#),
        ]
        return p
    }

    /// `-D` dynamic SOCKS proxy with compression, tunnels-only.
    private static var socksProxyExample: SSHProfile {
        var p = SSHProfile()
        p.name = "Example: SOCKS proxy (-D)"
        p.host = "gateway.example.com"
        p.username = "tunnel"
        p.openShell = false
        p.compression = true         // -C, handy over slow links
        p.keepAlive = true
        p.theme = "dracula"
        p.forwards = [
            PortForward(type: .dynamic, listenPort: "1080"),
        ]
        return p
    }

    /// `-J` jump host + private key + interactive shell + two local forwards + snippets.
    private static var bastionShellExample: SSHProfile {
        var p = SSHProfile()
        p.name = "Example: Bastion + shell (-J)"
        p.host = "10.0.5.20"                       // private host, only reachable via the bastion
        p.username = "admin"
        p.jumpHost = "jump@bastion.example.com"     // -J
        p.identityFile = "~/.ssh/id_ed25519"        // -i (a specific key)
        p.openShell = true                          // interactive shell alongside the tunnels
        p.keepAlive = true
        p.compression = true
        p.theme = "homebrew"
        p.extraOptions = "-o StrictHostKeyChecking=accept-new"
        p.forwards = [
            PortForward(type: .local,
                        listenPort: "8080",
                        targetHost: "127.0.0.1",
                        targetPort: "80"),          // web UI on the private box
            PortForward(type: .local,
                        listenPort: "6379",
                        targetHost: "127.0.0.1",
                        targetPort: "6379"),        // Redis
        ]
        p.snippets = [
            CommandSnippet(label: "Tail app log", command: "sudo journalctl -u myapp -f"),
            CommandSnippet(label: "Disk usage", command: "df -h"),
        ]
        return p
    }

    /// `-R` remote forward: expose this Mac's local dev server on the remote host.
    private static var remoteShareExample: SSHProfile {
        var p = SSHProfile()
        p.name = "Example: Share local port (-R)"
        p.host = "public.example.com"
        p.username = "web"
        p.openShell = false
        p.keepAlive = true
        p.theme = "novel"
        p.forwards = [
            PortForward(type: .remote,
                        listenPort: "9000",
                        targetHost: "127.0.0.1",
                        targetPort: "3000"),        // remote :9000 → your Mac's :3000
        ]
        return p
    }
}

File: cross-platform/src/RemoteStuff/Services/ExampleProfiles.cs
using System.Collections.Generic;
using RemoteStuff.Models;

namespace RemoteStuff.Services;

/// <summary>
/// Example profiles seeded on first launch so a new user can see how different
/// options come together. They use <c>example.com</c> placeholder hosts.
/// </summary>
public static class ExampleProfiles
{
    public static List<SshProfile> All() => new()
    {
        DatabaseTunnel(),
        SocksProxy(),
        BastionShell(),
        RemoteShare()
    };

    /// <summary>-L local forward, tunnels-only (-N): reach a remote database on localhost.</summary>
    private static SshProfile DatabaseTunnel() => new()
    {
        Name = "Example: Database tunnel (-L)",
        Host = "db.internal.example.com",
        Username = "deploy",
        OpenShell = false,
        KeepAlive = true,
        Theme = "ocean",
        Forwards = new List<PortForward>
        {
            new() { Type = ForwardType.Local, ListenPort = "5433", TargetHost = "127.0.0.1", TargetPort = "5432" }
        },
        Snippets = new List<CommandSnippet>
        {
            new() { Label = "Connect with psql", Command = "psql \"host=127.0.0.1 port=5433 dbname=app user=app\"" }
        }
    };

    /// <summary>-D dynamic SOCKS proxy with compression, tunnels-only.</summary>
    private static SshProfile SocksProxy() => new()
    {
        Name = "Example: SOCKS proxy (-D)",
        Host = "gateway.example.com",
        Username = "tunnel",
        OpenShell = false,
        Compression = true,
        KeepAlive = true,
        Theme = "dracula",
        Forwards = new List<PortForward>
        {
            new() { Type = ForwardType.Dynamic, ListenPort = "1080" }
        }
    };

    /// <summary>-J jump host + private key + interactive shell + local forwards + snippets.</summary>
    private static SshProfile BastionShell() => new()
    {
        Name = "Example: Bastion + shell (-J)",
        Host = "10.0.5.20",
        Username = "admin",
        JumpHost = "jump@bastion.example.com",
        IdentityFile = "~/.ssh/id_ed25519",
        OpenShell = true,
        KeepAlive = true,
        Compression = true,
        Theme = "homebrew",
        ExtraOptions = "-o StrictHostKeyChecking=accept-new",
        Forwards = new List<PortForward>
        {
            new() { Type = ForwardType.Local, Category = ForwardCategory.Webpage, ListenPort = "8080", TargetHost = "127.0.0.1", TargetPort = "80" },
            new() { Type = ForwardType.Local, Category = ForwardCategory.Redis, ServiceUsername = "default", ListenPort = "6379", TargetHost = "127.0.0.1", TargetPort = "6379" },
            new() { Type = ForwardType.Local, Category = ForwardCategory.Mqtt, ServiceUsername = "admin", ListenPort = "1883", TargetHost = "127.0.0.1", TargetPort = "1883" }
        },
        Snippets = new List<CommandSnippet>
        {
            new() { Label = "Tail app log", Command = "sudo journalctl -u myapp -f" },
            new() { Label = "Disk usage", Command = "df -h" }
        }
    };

    /// <summary>-R remote forward: expose this machine's local dev server on the remote host.</summary>
    private static SshProfile RemoteShare() => new()
    {
        Name = "Example: Share local port (-R)",
        Host = "public.example.com",
        Username = "web",
        OpenShell = false,
        KeepAlive = true,
        Theme = "novel",
        Forwards = new List<PortForward>
        {
            new() { Type = ForwardType.Remote, ListenPort = "9000", TargetHost = "127.0.0.1", TargetPort = "3000" }
        }
    };
}
